src/engine/loopingcontrol.cpp

```diff
@@ -367,157 +387,167 @@ void LoopingControl::hintReader(HintVector* pHintList) {
+        if (closestBeat != -1 && m_pBeats) {
+            pos = static_cast<int>(floor(m_pBeats->findNthBeat(closestBeat, -2)));
+            if (pos == -1 || (loopSamples.end - pos) < MINIMUM_AUDIBLE_LOOP_SIZE) {
+                pos = loopSamples.end - MINIMUM_AUDIBLE_LOOP_SIZE;
+            }
+        } else {
+            pos = loopSamples.end - MINIMUM_AUDIBLE_LOOP_SIZE;
         }
-//         qDebug() << "set loop_in to " << m_iLoopStartSample;
     }
+
+    if (pos != -1 && !even(pos)) {
+        pos--;
+    }
+
+    loopSamples.start = pos;
+    m_pCOLoopStartPosition->set(loopSamples.start);
+
+    m_loopSamples.setValue(loopSamples);
+    //qDebug() << "set loop_in to " << loopSamples.start;
 }
 
 void LoopingControl::slotLoopOut(double val) {
-    if (!m_pTrack) {
+    if (!m_pTrack || val <= 0.0) {
         return;
     }
-    if (val) {
-        int pos =
-                (m_pQuantizeEnabled->get() > 0.0 && m_pClosestBeat->get() != -1) ?
-                static_cast<int>(floor(m_pClosestBeat->get())) : m_iCurrentSample;
 
-        // If the user is trying to set a loop-out before the loop in or without
-        // having a loop-in, then ignore it.
-        if (m_iLoopStartSample == kNoTrigger || pos < m_iLoopStartSample) {
-            return;
+    LoopSamples loopSamples = m_loopSamples.getValue();
+    double closestBeat = -1;
+    int pos = m_iCurrentSample;
+    if (m_pQuantizeEnabled->toBool()) {
+        closestBeat = m_pClosestBeat->get();
+        if (closestBeat != -1) {
+            pos = static_cast<int>(floor(closestBeat));
         }
+    }
 
-        // If the loop-in and out points are set so close that the loop would be
-        //  inaudible (which can happen easily with quantize-to-beat enabled,)
-        //  use the smallest pre-defined beatloop instead (when possible)
-        if (pos - m_iLoopStartSample < MINIMUM_AUDIBLE_LOOP_SIZE) {
-            pos = m_iLoopStartSample;
-            if (m_pQuantizeEnabled->get() > 0.0 && m_pBeats) {
-                // 1 would have just returned loop_in, so give 2 to get the beat
-                // following loop_in
-                int nextbeat = m_pBeats->findNthBeat(m_iLoopStartSample, 2);
-                pos += (nextbeat - pos) * s_dBeatSizes[0];
-            } else {
-                pos += MINIMUM_AUDIBLE_LOOP_SIZE;
+    // If the user is trying to set a loop-out before the loop in or without
+    // having a loop-in, then ignore it.
+    if (loopSamples.start == kNoTrigger || pos < loopSamples.start) {
+        return;
+    }
+
+    // If the loop-in and out points are set so close that the loop would be
+    //  inaudible (which can happen easily with quantize-to-beat enabled,)
+    //  use the smallest pre-defined beatloop instead (when possible)
+    if ((pos - loopSamples.start) < MINIMUM_AUDIBLE_LOOP_SIZE) {
+        if (closestBeat != -1 && m_pBeats) {
+            pos = static_cast<int>(floor(m_pBeats->findNthBeat(closestBeat, 2)));
```

Comment:
The original code measured 1 beat length and then made the loop size s_dBeatSizes[0] times that length -- it looks to me like the new code would make the loop a whole beat long -- was that intended?